File: plugin/old/converter/js/plugin-modules/dependency-checker.js
/**
 * PIXLY 依赖Check工具module
 * 依赖: PIXLY命名空间, Node.js APIs
 */
(function(window) {
    'use strict';
    
    window.PIXLY = window.PIXLY || {};
    
    // Log instance for the module
    const log = window.pixlyLog;
    
    /**
     * 依赖Check器
     */
    const DependencyChecker = {
        
        /**
         * Checkall依赖
         */
        checkAll: async function() {
        if (log) {
            log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_CHECK_START, {}));
        }
            
            const results = {
                // pixly: await this.checkPixly(), // ❌ 已废弃 - 不存在的工具
                cjxl: await this.checkCjxl(),
                avifenc: await this.checkAvifenc(),
                cwebp: await this.checkCwebp(),
                ffmpeg: await this.checkFfmpeg(),
                exiftool: await this.checkExiftool()
            };
            
            const allInstalled = Object.values(results).every(r => r.installed);
            
        if (log) {
            log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_CHECK_COMPLETE, { status: allInstalled ? '✅' : '⚠️' }));
        }
            
            return {
                allInstalled,
                results
            };
        },
        
        /**
         * CheckPIXLY
         */
        checkPixly: async function() {
            return await this.checkCommand('pixly --version', 'PIXLY');
        },
        
        /**
         * CheckCJXL
         */
        checkCjxl: async function() {
            return await this.checkCommand('cjxl --version', 'JPEG XL (cjxl)');
        },
        
        /**
         * Checkavifenc
         */
        checkAvifenc: async function() {
            return await this.checkCommand('avifenc --version', 'AVIF Encoder');
        },
        
        /**
         * Checkcwebp
         */
        checkCwebp: async function() {
            return await this.checkCommand('cwebp -version', 'WebP Encoder');
        },
        
        /**
         * Checkffmpeg
         */
        checkFfmpeg: async function() {
            return await this.checkCommand('ffmpeg -version', 'FFmpeg');
        },
        
        /**
         * Checkexiftool
         */
        checkExiftool: async function() {
            return await this.checkCommand('exiftool -ver', 'ExifTool');
        },
        
        /**
         * Check命令is否available
         */
        checkCommand: async function(command, name) {
            try {
                const { exec } = require('child_process');
                const { promisify } = require('util');
                const execAsync = promisify(exec);
                
                // settings完整 PATH
                const env = Object.assign({}, require('process').env);
                env.PATH = `/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:${env.PATH || ''}`;
                
                const { stdout, stderr } = await execAsync(command, { env });
                const version = this.extractVersion(stdout || stderr);
                
                if (log) {
                    log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_FOUND, { name, version: version || 'installed' }));
                }
                
                return {
                    name,
                    installed: true,
                    version: version || 'unknown',
                    command: command.split(' ')[0]
                };
            } catch (error) {
                if (log) {
                    log.warn('PIXLY Dep', formatLog(LOG.DEPENDENCY_NOT_FOUND, { name }));
                }
                
                return {
                    name,
                    installed: false,
                    version: null,
                    command: command.split(' ')[0],
                    error: error.message
                };
            }
        },
        
        /**
         * from输出中提取version号
         */
        extractVersion: function(output) {
            if (!output) return null;
            
            // 匹配常见version号格式
            const patterns = [
                /version\s+(\d+\.\d+\.\d+)/i,
                /v(\d+\.\d+\.\d+)/i,
                /(\d+\.\d+\.\d+)/,
                /(\d+\.\d+)/
            ];
            
            for (const pattern of patterns) {
                const match = output.match(pattern);
                if (match && match[1]) {
                    return match[1];
                }
            }
            
            return null;
        },
        
        /**
         * 获取installed建议
         */
        getInstallSuggestions: function(missingDeps) {
            const suggestions = {
                // 'pixly': 'brew install pixly', // ❌ 已废弃 - 不存在的工具
                'cjxl': 'brew install jpeg-xl',
                'avifenc': 'brew install libavif',
                'cwebp': 'brew install webp',
                'ffmpeg': 'brew install ffmpeg',
                'exiftool': 'brew install exiftool'
            };
            
            return missingDeps.map(dep => ({
                name: dep.name,
                command: suggestions[dep.command] || `brew install ${dep.command}`
            }));
        },
        
        /**
         * 自动installed缺失依赖
         */
        autoInstall: async function(missingDeps) {
        if (log) {
            log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_AUTO_INSTALL_START, {}));
        }
            
            const { exec } = require('child_process');
            const { promisify } = require('util');
            const execAsync = promisify(exec);
            
            const results = [];
            
            for (const dep of missingDeps) {
                const suggestions = this.getInstallSuggestions([dep]);
                const installCmd = suggestions[0].command;
                
                try {
                    if (log) {
                        log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_INSTALLING, { name: dep.name }));
                    }
                    await execAsync(installCmd);
                    
                    results.push({
                        name: dep.name,
                        success: true
                    });
                    
                    if (log) {
                        log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_INSTALL_SUCCESS, { name: dep.name }));
                    }
                } catch (error) {
                    results.push({
                        name: dep.name,
                        success: false,
                        error: error.message
                    });
                    
                if (log) {
                    log.error('PIXLY Dep', formatLog(LOG.DEPENDENCY_INSTALL_FAILED, { name: dep.name, error: error.message || error }));
                }
                }
            }
            
            return results;
        },
        
        /**
         * 生成依赖报告
         */
        generateReport: function(checkResults) {
            const report = [];
            
            report.push('='.repeat(50));
            report.push('PIXLY 依赖Check报告');
            report.push('='.repeat(50));
            report.push('');
            
            for (const [key, result] of Object.entries(checkResults.results)) {
                const status = result.installed ? '✅' : '❌';
                const version = result.version || 'N/A';
                report.push(`${status} ${result.name.padEnd(20)} ${version}`);
            }
            
            report.push('');
            report.push('='.repeat(50));
            report.push(`Overall Status: ${checkResults.allInstalled ? '✅ All dependencies installed' : '⚠️ Missing some dependencies'}`);
            report.push('='.repeat(50));
            
            return report.join('\n');
        }
    };
    
    // 暴露to全局
    PIXLY.DependencyChecker = DependencyChecker;
    
    // 向后兼容
    window.checkAllDependencies = function() {
        return DependencyChecker.checkAll();
    };
    
    // ❌ 已废弃 - Pixly不是独立CLI工具
    // window.checkPixlyQuietly = function() {
    //     return DependencyChecker.checkPixly().then(r => r.installed);
    // };
    
    window.autoInstallToolsSilently = function() {
        return DependencyChecker.checkAll().then(results => {
            const missing = Object.values(results.results).filter(r => !r.installed);
            if (missing.length > 0) {
                return DependencyChecker.autoInstall(missing);
            }
            return [];
        });
    };
    
    if (log) {
        log.info('PIXLY Dep', formatLog(LOG.DEPENDENCY_CHECKER_MODULE_LOADED, {}));
    }
    
})(window);

File: src/dependency_checker.rs
/**
 * Dependency Checker - 外部依赖检查
 * 
 * 检查必需的外部工具：
 * - exiftool (元数据处理)
 * - ffmpeg (视频处理)
 * - ffprobe (视频分析)
 * - cjxl/djxl (JPEG XL)
 * - avifenc/avifdec (AVIF)
 */
use std::process::Command;
use std::path::PathBuf;
use std::env;
use log::{debug, info};

/// 依赖项信息
#[derive(Debug, Clone)]
pub struct Dependency {
    pub name: &'static str,
    pub required: bool,
    pub description: &'static str,
    pub install_hint: &'static str,
}

/// 检查结果
#[derive(Debug)]
pub struct CheckResult {
    pub name: &'static str,
    pub available: bool,
    pub version: Option<String>,
}

impl Dependency {
    /// 检查依赖是否可用
    pub fn check(&self) -> CheckResult {
        let (available, version) = match self.name {
            "ffmpeg" => {
                if let Some(path) = find_ffmpeg() {
                    (true, get_command_version_by_path(&path))
                } else {
                    (false, None)
                }
            }
            "ffprobe" => {
                if let Some(ffmpeg_path) = find_ffmpeg() {
                    if let Some(parent) = ffmpeg_path.parent() {
                        let ffprobe_path = parent.join("ffprobe");
                        if ffprobe_path.exists() {
                            (true, get_command_version_by_path(&ffprobe_path))
                        } else {
                            (is_command_available("ffprobe"), get_command_version("ffprobe"))
                        }
                    } else {
                        (is_command_available("ffprobe"), get_command_version("ffprobe"))
                    }
                } else {
                    (is_command_available("ffprobe"), get_command_version("ffprobe"))
                }
            }
            "exiftool" => {
                if let Some(path) = find_exiftool() {
                    (true, get_command_version_by_path(&path))
                } else {
                    (false, None)
                }
            }
            _ => {
                let available = is_command_available(self.name);
                let version = if available { get_command_version(self.name) } else { None };
                (available, version)
            }
        };
        
        CheckResult { name: self.name, available, version }
    }
}

fn is_command_available(cmd: &str) -> bool {
    Command::new(cmd)
        .arg("--version")
        .output()
        .map(|output| output.status.success())
        .unwrap_or(false)
}

fn get_command_version(cmd: &str) -> Option<String> {
    Command::new(cmd)
        .arg("--version")
        .output()
        .ok()
        .and_then(|output| {
            if output.status.success() {
                String::from_utf8(output.stdout).ok()
                    .and_then(|s| s.lines().next().map(|l| l.to_string()))
            } else {
                None
            }
        })
}

fn get_command_version_by_path(path: &PathBuf) -> Option<String> {
    let path_str = path.to_string_lossy();
    let args = if path_str.contains("exiftool") {
        vec!["-ver"]
    } else if path_str.contains("ffmpeg") || path_str.contains("ffprobe") {
        vec!["-version"]
    } else {
        vec!["--version"]
    };
    
    Command::new(path)
        .args(&args)
        .output()
        .ok()
        .and_then(|output| {
            if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
                let first_line = stdout.lines().next().unwrap_or("").trim();
                if !first_line.is_empty() {
                    if path_str.contains("exiftool") && !first_line.contains("exiftool") {
                        return Some(format!("ExifTool {}", first_line));
                    }
                    return Some(first_line.to_string());
                }
            }
            if let Ok(stderr) = String::from_utf8(output.stderr) {
                let first_line = stderr.lines().next().unwrap_or("").trim();
                if !first_line.is_empty() {
                    return Some(first_line.to_string());
                }
            }
            None
        })
}

pub fn find_ffmpeg() -> Option<PathBuf> {
    if let Ok(path) = env::var("FFMPEG_PATH") {
        let ffmpeg_path = PathBuf::from(path);
        if ffmpeg_path.exists() {
            return Some(ffmpeg_path);
        }
    }
    
    let mut common_paths: Vec<String> = Vec::new();
    match env::consts::OS {
        "macos" => {
            common_paths.push("/opt/homebrew/bin/ffmpeg".to_string());
            common_paths.push("/usr/local/bin/ffmpeg".to_string());
            common_paths.push("/opt/local/bin/ffmpeg".to_string());
        }
        "windows" => {
            common_paths.push("C:\\Program Files\\FFmpeg\\bin\\ffmpeg.exe".to_string());
            common_paths.push("C:\\ffmpeg\\bin\\ffmpeg.exe".to_string());
            if let Ok(userprofile) = env::var("USERPROFILE") {
                common_paths.push(format!("{}\\ffmpeg\\bin\\ffmpeg.exe", userprofile));
            }
        }
        "linux" => {
            common_paths.push("/usr/bin/ffmpeg".to_string());
            common_paths.push("/usr/local/bin/ffmpeg".to_string());
            common_paths.push("/opt/ffmpeg/bin/ffmpeg".to_string());
            common_paths.push("/snap/bin/ffmpeg".to_string());
        }
        _ => {}
    }
    
    for path_str in common_paths {
        let path = PathBuf::from(&path_str);
        if path.exists() {
            return Some(path);
        }
    }
    
    if is_command_available("ffmpeg") {
        #[cfg(target_os = "windows")]
        let which_cmd = "where";
        #[cfg(not(target_os = "windows"))]
        let which_cmd = "which";
        
        if let Ok(output) = Command::new(which_cmd).arg("ffmpeg").output()
            && output.status.success()
            && let Ok(path_str) = String::from_utf8(output.stdout)
        {
            return Some(PathBuf::from(path_str.trim()));
        }
        debug!("⚠️  FFmpeg not found in PATH");
    }
    None
}

pub fn find_exiftool() -> Option<PathBuf> {
    if let Ok(path) = env::var("EXIFTOOL_PATH") {
        let exiftool_path = PathBuf::from(path);
        if exiftool_path.exists() {
            return Some(exiftool_path);
        }
    }
    
    let mut common_paths: Vec<String> = Vec::new();
    match env::consts::OS {
        "macos" => {
            common_paths.push("/opt/homebrew/bin/exiftool".to_string());
            common_paths.push("/usr/local/bin/exiftool".to_string());
        }
        "windows" => {
            common_paths.push("C:\\exiftool\\exiftool.exe".to_string());
            if let Ok(userprofile) = env::var("USERPROFILE") {
                common_paths.push(format!("{}\\exiftool\\exiftool.exe", userprofile));
            }
        }
        "linux" => {
            common_paths.push("/usr/bin/exiftool".to_string());
            common_paths.push("/usr/local/bin/exiftool".to_string());
        }
        _ => {}
    }
    
    for path_str in common_paths {
        let path = PathBuf::from(&path_str);
        if path.exists() {
            return Some(path);
        }
    }
    
    if is_command_available("exiftool") {
        return Some(PathBuf::from("exiftool"));
    }
    None
}

pub fn get_dependencies() -> Vec<Dependency> {
    vec![
        Dependency {
            name: "exiftool",
            required: true,
            description: "元数据处理（EXIF/XMP/ICC）",
            install_hint: get_install_hint("exiftool"),
        },
        Dependency {
            name: "ffmpeg",
            required: false,
            description: "视频转换和GIF分析",
            install_hint: get_install_hint("ffmpeg"),
        },
        Dependency {
            name: "ffprobe",
            required: false,
            description: "视频信息获取",
            install_hint: get_install_hint("ffmpeg"),
        },
        Dependency {
            name: "cjxl",
            required: false,
            description: "JPEG XL编码",
            install_hint: get_install_hint("cjxl"),
        },
        Dependency {
            name: "djxl",
            required: false,
            description: "JPEG XL解码",
            install_hint: get_install_hint("cjxl"),
        },
        Dependency {
            name: "avifenc",
            required: false,
            description: "AVIF编码",
            install_hint: get_install_hint("avifenc"),
        },
        Dependency {
            name: "avifdec",
            required: false,
            description: "AVIF解码",
            install_hint: get_install_hint("avifenc"),
        },
    ]
}

fn get_install_hint(tool: &str) -> &'static str {
    match env::consts::OS {
        "macos" => match tool {
            "exiftool" => "macOS: brew install exiftool",
            "ffmpeg" => "macOS: brew install ffmpeg",
            "cjxl" => "macOS: brew install jpeg-xl",
            "avifenc" => "macOS: brew install libavif",
            _ => "请查看官方文档",
        },
        "windows" => match tool {
            "exiftool" => "Windows: 下载 https://exiftool.org/",
            "ffmpeg" => "Windows: 下载 https://ffmpeg.org/",
            "cjxl" => "Windows: 下载 https://github.com/libjxl/libjxl/releases",
            "avifenc" => "Windows: 下载 https://github.com/AOMediaCodec/libavif/releases",
            _ => "请查看官方文档",
        },
        "linux" => match tool {
            "exiftool" => "Linux: sudo apt install libimage-exiftool-perl",
            "ffmpeg" => "Linux: sudo apt install ffmpeg",
            "cjxl" => "Linux: sudo apt install libjxl-tools",
            "avifenc" => "Linux: sudo apt install libavif-bin",
            _ => "请查看官方文档",
        },
        _ => "请查看官方文档",
    }
}

pub fn check_all_dependencies() -> Vec<CheckResult> {
    get_dependencies().iter().map(|dep| dep.check()).collect()
}

pub fn check_and_report_dependencies(verbose: bool) -> bool {
    let dependencies = get_dependencies();
    let results = check_all_dependencies();
    
    let mut all_required_available = true;
    let mut warnings = Vec::new();
    
    if verbose {
        info!("🔍 Checking external dependencies...");
    }
    
    for (dep, result) in dependencies.iter().zip(results.iter()) {
        if result.available {
            if verbose {
                if let Some(version) = &result.version {
                    info!("  ✅ {} - {} ({})", 
                        result.name, dep.description,
                        version.lines().next().unwrap_or("unknown"));
                } else {
                    info!("  ✅ {} - {}", result.name, dep.description);
                }
            }
        } else if dep.required {
            all_required_available = false;
            eprintln!("  ❌ {} - {} (required)", result.name, dep.description);
            eprintln!("     {}", dep.install_hint);
        } else {
            warnings.push(result.name);
            if verbose {
                info!("  ⚠️  {} - {} (optional)", result.name, dep.description);
            }
        }
    }
    
    if !warnings.is_empty() && verbose {
        info!("💡 Optional tools not installed: {}", warnings.join(", "));
    }
    
    if !all_required_available {
        eprintln!("❌ Missing required dependencies!");
    } else if verbose {
        info!("✅ All required dependencies are satisfied!");
    }
    
    all_required_available
}

pub fn check_required_dependencies_quiet() -> bool {
    get_dependencies()
        .iter()
        .filter(|dep| dep.required)
        .all(|dep| is_command_available(dep.name))
}
